### DISCUSSION TITLE
How can I access session data in middleware?

### DISCUSSION THREAD

**Discussion by ipimpat** [ckan/ckan#6520] [Q&A] [answered]
Sorry for my newbie question, but I'm a Python novice, and no experience with Flask.

I'm trying to make a middleware plugin that prevents anonymous access to my CKAN site, but I'm also using the ldap extension, so I need to check if the ckanext-ldap-user session key is set to determine if the user is logged in.

Gist: https://gist.github.com/ipimpat/5b9614db4c1e4794cc5d50a347e4a8ef

How ever I get this error when trying to access the session:

`an          | TypeError: No object (name: session) has been registered for this thread`

In my code I have this method which should check if the ckanext-ldap-user session key exists.

In the top of my code I'm importing the session variable
```python
from ckan.common import config, session
```

And then try to use it like this

```python
    def _is_logged_in(self, environ):
        # check if the request belongs to an authenticated session
        if 'ckanext-ldap-user' in session:        # <-- this breaks
            return True
        elif 'repoze.who.identity' in environ:
            return True

        return False
```

**Comment by frafra** [comment, marked as answer]
`session` is not defined, as the error says. You just have `self` and `environ` available.

**Reply by ipimpat** [reply]
> `session` is not defined, as the error says.

Updated my question - I do import it, sorry I did not mention that.

**Reply by smotornyuk** [reply]
`session` is a request-specific variable that is initialized by flask for the functions that handle requests.  There are other similar variables in CKAN common, like `request`, `g`. 
You should never use them outside the view functions/blueprints.

Middleware it's a wrapper **around** application, not just some functions that are called before the request and after the request. That's why you should not use tools from the inside of application as a part of your middleware

Either move the logic of your middleware to a different place(Flask's `app.before_request` hook) or use only the `environ` variable(headers, cookies) inside `__call__` method of your middleware.
PS there is a way to use sessions inside middleware, but it's a really bad thing to do, so I prefer not to share this knowledge :laughing:

**Reply by ipimpat** [reply]
> `session` is a request-specific variable that is initialized by flask for the functions that handle requests. There are other similar variables in CKAN common, like `request`, `g`.
> You should never use them outside the view functions/blueprints.

Thanks for the advice, as I mentioned in the beginning of my question, I'm a Python novice, so I'm ignorant about that stuff (I have some experience with Laravel/PHP).

> Either move the logic of your middleware to a different place(Flask's `app.before_request` hook)

Could you provide a bit of boiler plate code to get me startet - just the "entry point" part ?

> PS there is a way to use sessions inside middleware, but it's a really bad thing to do, so I prefer not to share this knowledge laughing

I must admit I'm curious to see it, to better understand why it's a bad idea.

**Reply by ipimpat** [reply]
> > Either move the logic of your middleware to a different place(Flask's `app.before_request` hook)
> 
> Could you provide a bit of boiler plate code to get me startet - just the "entry point" part ?

Never mind, I think I worked it out, is this what you are thinking about?

```python

    def __call__(self, environ, start_response):

        @self.app.before_request
        def before_request_callback():
            log.info(session)   # <--  works
```

**Reply by smotornyuk** [reply]
Yep, exactly

And, as for a fast dirty way to use sessions inside middleware, you can rewrite it to something like:

```python
class Middleware:
    session = None

    def __init__(self, app, app_conf):
        import ckan.common
        app.before_request(lambda: setattr(self, 'session', dict(ckan.common.session))


    def __call__(self, environ, start_response):
         log.info(self.session)
```

In that way, you are copying session data into middleware for each request. But as it's only a copy and we have to use `before_request` anyway, it's better to do as per your last comment
